hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java

```diff
@@ -0,0 +1,271 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.io.compress.zlib;
+
+import java.io.IOException;
+import java.util.zip.Checksum;
+import java.util.zip.Deflater;
+import java.util.zip.GZIPOutputStream;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.compress.Compressor;
+import org.apache.hadoop.io.compress.DoNotPool;
+import org.apache.hadoop.util.DataChecksum;
+
+/**
+ * A {@link Compressor} based on the popular gzip compressed file format.
+ * http://www.gzip.org/
+ */
+@DoNotPool
+public class BuiltInGzipCompressor implements Compressor {
+
+  /**
+   * Fixed ten-byte gzip header. See {@link GZIPOutputStream}'s source for
+   * details.
+   */
+  private static final byte[] GZIP_HEADER = new byte[]{
+      0x1f, (byte) 0x8b, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+  // The trailer will be overwritten based on crc and output size.
+  private static final byte[] GZIP_TRAILER = new byte[]{
+      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+  private static final int GZIP_HEADER_LEN = GZIP_HEADER.length;
+  private static final int GZIP_TRAILER_LEN = GZIP_TRAILER.length;
+
+  private Deflater deflater;
+
+  private boolean finished;
+
+  private int headerOff = 0;
+  private int trailerOff = 0;
+
+  private int numExtraBytesWritten = 0;
+
+  private int currentBufLen = 0;
+  private int accuButLen = 0;
+
+  private final Checksum crc = DataChecksum.newCrc32();
+
+  private BuiltInGzipDecompressor.GzipStateLabel state;
+
+  public BuiltInGzipCompressor(Configuration conf) {
+    init(conf);
+  }
+
+  @Override
+  public boolean finished() {
+    // Only if the trailer is also written, it is thought as finished.
+    return finished && state == BuiltInGzipDecompressor.GzipStateLabel.FINISHED;
+  }
+
+  @Override
+  public boolean needsInput() {
+    if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+      return deflater.needsInput();
+    }
+
+    // After we output the trailer for the current input, we can take another input.
```

Comment:
This comment is outdated. Also I wonder if we can change it to something like:
```java
    if (!deflater.needsInput()) {
      return false;
    }

    // After we output the trailer for the current input, we can take another input.
    return state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM;
```

It seems strange that we'd need more input when state is `FINISHED`.

Reply:
Then `needsInput` will return false for `HEADER_BASIC` state? Seems no correct.

Reply:
Seems to be:

```java
if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
  return deflater.needsInput();
}

return state == BuiltInGzipDecompressor.GzipStateLabel.HEADER_BASIC;
```

Reply:
In `HEADER_BASIC` it'll return false since `deflater.needsInput` will return false (since it hasn't start processing the input yet). I think this logic is correct since we need it to go into `compress` method?

Reply:
Hm, but `deflater.needsInput()` actually return true after we just create Deflater instance (not call `setInput` yet).

Reply:
Yes that's right. The usage pattern for this seems to be:
```java
    compressor.setInput(b, off, len);
    while (!compressor.needsInput()) {
      compress();
    }
```
so `setInput` is always called first. Although, it might be safer to also add a check of `HEADER_BASIC` there too.

In the code snippet you have above, seems it will return true when for the first time `compressor.setInput` is called (since the state is `HEADER_BASIC`) , and it will not call `compress` afterwards?

Reply:
Rethinking about it and testing locally, I think it should be 

```java
deflater.needsInput() && state !=BuiltInGzipDecompressor.GzipStateLabel.TRAILER_CRC;
```

Reply:
We should return `deflater.needsInput()` in most cases. But if we are writing the trailer, even it needs input, we should return false.